python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,251 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Utility module for various conversions around PrimitiveType implementations
+
+This module enables:
+    - Converting partition strings to built-in python objects
+    - Converting a value to a byte buffer
+    - Converting a byte buffer to a value
+
+Note:
+    Conversion logic varies based on the PrimitiveType implementation. Therefore conversion functions
+    are defined here as generic functions using the @singledispatch decorator. For each PrimitiveType
+    implementation, a concrete function is registered for each generic conversion function. For PrimitiveType
+    implementations that share the same conversion logic, registrations can be stacked.
+
+Example (registering `from_partition_to_py` logic for BooleanType):
+    >>> @from_partition_value_to_py.register(BooleanType)
+    >>> def _(type_, partition_value_as_str: str):
+            if partition_value_as_str is None:
+                return False
+            return partition_value_as_str.lower() == "true"
+
+Example (stacking registrations):
+   >>> @from_partition_value_to_py.register(FloatType)
+   >>> @from_partition_value_to_py.register(DoubleType)
+   >>> def _(type_, partition_value_as_str: str):
+           return float(partition_value_as_str)
+"""
+import struct
+import uuid
+from decimal import ROUND_HALF_UP, Decimal
+from functools import singledispatch
+from typing import Any
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+HIVE_NULL = "__HIVE_DEFAULT_PARTITION__"
+
+
+@singledispatch
+def from_partition_value_to_py(type_, partition_value_as_str: str):
```

Comment:
BTW, also is it feasible to put type annotation for `type_`?

Reply:
Agreed, I'll rename `type_` to `primitive_type`, that's much clearer. For adding the type annotation, this should be feasible in that I don't think it will break the single dispatch decorator. Are you thinking we hint using the base primitive type class like `def from_partition_value_to_py(type_: PrimitiveType, partition_value_as_str: str):`?

Reply:
Updated! Thanks @jun-he, this looks much cleaner.